firmware/arduino/chessboard/uart.ino

```diff
@@ -58,54 +67,91 @@ void serialEvent2()
+            // Returns true for valid input and false for invalid input.
             currentState = EXECUTING;
-            if (parseMessageFromPi(buffer))
-                errorCode = NONE;
+            if (validateMessageFromPi(buffer))
+                errorCode = NO_ERROR;
+
 
-            // Send response
+            // Sends acknowledgement and any errors with the instruction
             sendMessageToPi(currentState, buffer[5], errorCode);
+
+            if (errorCode == NO_ERROR)
+            {
+                // Makes move, returns success
+                if(makeMove(buffer))
+                    currentState = IDLE;
+
+                sendMessageToPi(currentState, buffer[5], errorCode);
+            }
         }
     }
 }
 
-bool parseMessageFromPi(char * message)
+// Check that the instruction is valid
+bool validateMessageFromPi(char * message)
 {
-    // Move types 0 and 1 are valid for the same range of values
-    if (message[0]  ==  DIRECT  ||  message[0]  ==  EDGES)
+    if (message[0] == DIRECT || message[0] == EDGES)
     {
-        // Check that the inputs are in a valid range
-        if (message[1] < 'A'  ||  message[1] > 'L' 
-           ||  message[2] < 'A'  ||  message[2] > 'L' 
-           ||  message[3] < 'A'  ||  message[3] > 'L' 
-           ||  message[4] < 'A'  ||   message[4] > 'L') 
+        if (isInvalidCoord(message[1]) || isInvalidCoord(message[2]) 
+            || isInvalidCoord(message[3]) || isInvalidCoord(message[4]))
         {
             errorCode = INVALID_LOCATION;
             currentState = ERROR;
             return false;
         }
-
-        // Move type 0
-        if (message[0]  ==  DIRECT)
-            moveDirect(message[1] - 'A', message[2] - 'A', message[3] - 'A', message[4] - 'A');
-        // Move type 1
-        else if (message[0] == EDGES)
-            moveAlongEdges(message[1] - 'A', message[2] - 'A', message[3] - 'A', message[4] - 'A');
-
     }
     else if (message[0] == ALIGN)
     {
-        // Check that the inputs are in a valid range
-        if (message[1] < 'A'  ||  message[1] > 'L' 
-           ||  message[2] < 'A'  ||  message[2] > 'L') 
+        if (isInvalidCoord(message[1]) || isInvalidCoord(message[2]))
         {
             errorCode = INVALID_LOCATION;
             currentState = ERROR;
             return false;
         }
+    }
+    else
+    {
+        // Invalid opcode
+        errorCode = INVALID_OP;
+        currentState = ERROR;
+        return false;
+    }
+    errorCode = NO_ERROR;
+    return true;
+}
 
-        // Center piece moved by player
-        alignPiece(message[1] - 'A', message[2] - 'A');
+bool makeMove(char * message)
+{
+    // Move type 0
+    if (message[0]  ==  DIRECT)
+    {
+        if (!moveDirect(message[2] - 'A', message[1] - 'A', message[4] - 'A', message[3] - 'A'))
+        {
+            currentState = ERROR;
+            errorCode = MOVEMENT_ERROR;
+            return false;
```

Comment:
You can minimize code duplication here (if you want)
```c
// At start of function
bool moveSucceeded = false;
...
// For each movement case
moveSucceeded = moveX(...);
...
// After last else condition
if (!moveSucceeded) {
    currentState = ERROR;
    errorCode = MOVEMENT_ERROR;
    return false;
}
...
```
You don't have to do this, just a suggestion